design/continuous-delivery.md

```diff
@@ -0,0 +1,329 @@
+# Continuous Delivery for CDK Apps
+
+This is a specification for continuous delivery support for CDK apps. It
+describes the requirements, proposed APIs and developer workflow.
+
+**_Goal_: full CI/CD for CDK apps at any complexity.**
+
+The desired developer experience is that teams will be able to “git push” a
+change into their CDK app repo and have this change automatically picked up,
+built, tested and deployed according to a deployment flow they define.
+
+Any changes in resources, assets *or stacks* in their apps will automatically be
+added. To that end, the deployment pipeline itself will also be continuously
+delivered and updated throughout the same workflow. No manual deployments to
+production (incl. the pipeline) will be required.
+
+The only caveat is that **new environments** (account/region) will need to be
+bootstrapped in advance in order to establish trust with the central pipeline
+environment and set-up resources needed for deployment such as the assets S3
+bucket.
+
+Prototype: https://github.com/eladb/cdkcd-test
+
+## Requirements
+
+This list describes only the minimal set of requirements from this feature.
+After we release these building blocks, we will look into vending higher-level
+"one liner" APIs that will make it very easy to get started.
+
+1. **Assets**: Support apps that include all supported kinds of assets (S3
+   files, ECR images)
+2. **Multiple-environments**: Support apps that have stacks that target multiple
+   environments (accounts/regions)
+3. **Orchestration**: Allow developers to express complex deployment
+   orchestration based on the capabilities of CodePipeline
+4. **User defined build + synth environment**: the runtime environment in which
+   the code is built and the CDK app is synthesized should be fully customizable
+   by the user
+5. **Controlled deployment environment**: for security reasons, the runtime
+   environment in which deployment is executed will be fully controlled and will
+   not allow running user code or user-defined image
+
```

Comment:
Maybe have costs in mind as well. For example, if you deploy the cdk stack in a CodeBuild action it costs you more than using the CloudFormation deploy action.

Reply:
Yes, this will be addressed by using the stock cloudformation actions. Stay tuned.